File: ProgramacaoDinamica/Java/Fibonacci.java
public class Fibonacci {
	/*
	 * mem[0] = 0
	 * mem[1] = 1
	 * mem[n]=mem[n-1] + mem[n-2]
	 */
	public static long[] mem = null;

	public static void main(String[] args) {
		int n = 40;
		long t = System.currentTimeMillis();
		long f = fibR(n);
		t = (System.currentTimeMillis() - t);
		System.out.format("Fibo: %d : Tempo: %d ms \n", f, t);
	}
	public static long fibI(long n){
		if(n==0) return 0;
		if(n==1) return 1;
		    
		long a = 0;
		long b = 1;
		long c = a+b;
		for(int i = 0; i<n-1; i++){
			c = a + b;
			a = b;
			b = c;
		}
		return c;
	}
	public static long fibR(int n) {
		//if (mem == null) mem = new long[n + 1]; 
		//if (mem[n] != 0) return mem[n]; //memoizacao
		if (n <= 0) return 0;
		if (n == 1) return 1;

		long l1 = fibR(n - 1);
		//mem[n - 1] = l1;
		long l2 = fibR(n - 2);
		//mem[n - 2] = l2;
		
		return l1 + l2;
	}
}

File: ProgramacaoDinamica/Python/Fibonacci.py
import sys
import time

class Fibonacci:
	"""
	 * mem[0] = 0
	 * mem[1] = 1
	 * mem[n]=mem[n-1] + mem[n-2]
	"""
	def __init__(self, n: int):
		self.mem = [0]*(n + 1)
		self.mem[0] = 0
		self.mem[1] = 1
		self.n = n

	def fibI(self):
		if(self.n <= 1): 
			return self.n
		res = 1
		ultimo = 1
		penultimo = 1
		for i in range(3, self.n+1):
			res = ultimo + penultimo
			penultimo = ultimo
			ultimo = res
		return res

	def fibRMem(self): return self.__fibRMem(self.n)
	def __fibRMem(self, n: int):
		if (n <=1): 
			return n
		if (self.mem[n] != 0): 
			return self.mem[n] #memoizacao

		ultimo = self.__fibRMem(n - 1)
		self.mem[n - 1] = ultimo
		penultimo = self.__fibRMem(n - 2)
		self.mem[n - 2] = penultimo
		return ultimo + penultimo

	def fibR(self): return self.__fibR(self.n)
	def __fibR(self, n: int):
		if (n <= 1): 
			return n
		return self.__fibR(n - 1) + self.__fibR(n - 2)

if __name__ == "__main__":
	n = 30
	if len(sys.argv) > 1:
		n = int(sys.argv[1])

	print("%s %20s %20s %20s" % ("N", "Resultado", "Tipo", "Tempo (ms)"))
	app = Fibonacci(n)
	inicio = time.time()*1000
	f = app.fibI()
	tempo = time.time()*1000 - inicio
	print("%d %20d %20s %20.2f ms" % (n, f, "Iterativo", tempo))

	inicio = time.time()*1000
	f = app.fibRMem()
	tempo = time.time()*1000 - inicio
	print("%d %20d %20s %20.2f ms" % (n, f, "Memoizado", tempo))

	inicio = time.time()*1000
	f = app.fibR()
	tempo = time.time()*1000 - inicio
	print("%d %20d %20s %20.2f ms" % (n, f, "Recursivo", tempo))

